lib/rules/attributes-order.js

```diff
@@ -0,0 +1,231 @@
+import createErrorMessage from '../helpers/create-error-message.js';
+import isValidConfigObjectFormat from '../helpers/is-valid-config-object.js';
+import Rule from './_base.js';
+
+const ERROR_MESSAGE = 'Your config does not match the allowed values';
+
+const AttributeType = {
+  ARGUMENTS: 'arguments',
+  ATTRIBUTES: 'attributes',
+  MODIFIERS: 'modifiers',
+  SPLATTRIBUTES: 'splattributes',
+};
+
+const DEFAULT_CONFIG = {
+  alphabetize: true,
+  attributeOrder: [
+    AttributeType.ARGUMENTS,
+    AttributeType.ATTRIBUTES,
+    AttributeType.MODIFIERS,
+    AttributeType.SPLATTRIBUTES,
+  ],
+};
+
+export function createAttributesOrderErrorMessage(config) {
+  return createErrorMessage('attributes-order', ERROR_MESSAGE, config);
+}
+
+export default class AttributesOrder extends Rule {
+  parseConfig(config) {
+    const errorMessage = createAttributesOrderErrorMessage(config);
+    if (!isValidConfigObjectFormat(config, DEFAULT_CONFIG)) {
+      throw new Error(errorMessage);
+    }
+
+    if (config && typeof config === 'object') {
+      return { ...DEFAULT_CONFIG, ...config };
+    }
+
+    return DEFAULT_CONFIG;
+  }
+
+  visitor() {
+    const log = this.log.bind(this);
+    const sourceForNode = this.sourceForNode.bind(this);
+    const config = this.config;
+
+    function getMustacheTokens(node, nodeText) {
+      const keys = [];
+      for (const attr of node.hash.pairs) {
+        const source = sourceForNode(attr);
+        keys.push([AttributeType.ATTRIBUTES, nodeText.indexOf(source), source, attr]);
+      }
+      return keys;
+    }
+
+    function getTokens(node, nodeText) {
+      const keys = [];
+      for (const attr of node.attributes) {
+        const source = sourceForNode(attr);
+        const type = attr.name.startsWith('@')
+          ? AttributeType.ARGUMENTS
+          : attr.name.startsWith('...')
+          ? AttributeType.SPLATTRIBUTES
+          : AttributeType.ATTRIBUTES;
+        keys.push([type, nodeText.indexOf(source), source, attr]);
+      }
+      for (const modifier of node.modifiers) {
+        const source = sourceForNode(modifier);
+        keys.push([AttributeType.MODIFIERS, nodeText.indexOf(source), source, modifier]);
+      }
+      return keys;
+    }
+
+    function groupTokens(tokens) {
```

Comment:
Can you add jsdoc comments to these functions (at least the non-trivial ones) explaining what they are for and details about the parameters / return types?